camayoc/data_provider.py

```diff
@@ -0,0 +1,116 @@
+        self._created_models = {}
+
+    def _select_definitions(self, match_criteria):
+        matching = self._defined_models.where(**match_criteria)
+        return list(matching)
+
+    def _create_model(self, model):
+        # FIXME: what should we do with possible HTTPError ?
+        model.create()
+        self._created_models[model.name] = model
+
+    def _create_dependencies(self, definition, new):
+        dependencies_ids = []
+        if issubclass(self._model_class, Credential):
+            return dependencies_ids
+
+        if issubclass(self._model_class, Source):
+            dependency_definitions = self._data_provider.credentials._select_definitions(
+                match_criteria={"name": Table.is_in(definition.credentials)}
+            )
+            for dependency_definition in dependency_definitions:
+                if new:
+                    dependency_definition = replace_definition_name(dependency_definition)
+                credential = self._data_provider.credentials._create_from_definition(
+                    definition=dependency_definition, new_dependencies=new, data_only=False
+                )
+                dependencies_ids.append(credential._id)
+            return dependencies_ids
+
+        if issubclass(self._model_class, Scan):
+            dependency_definitions = self._data_provider.sources._select_definitions(
+                match_criteria={"name": Table.is_in(definition.sources)}
+            )
+            for dependency_definition in dependency_definitions:
+                if new:
+                    dependency_definition = replace_definition_name(dependency_definition)
+                source = self._data_provider.sources._create_from_definition(
+                    definition=dependency_definition, new_dependencies=new, data_only=False
+                )
+                dependencies_ids.append(source._id)
+            return dependencies_ids
+
+    def _create_from_definition(self, definition, new_dependencies, data_only):
+        if existing_model := self._created_models.get(definition.name):
+            return existing_model
+
+        dependencies_ids = self._create_dependencies(definition, new=new_dependencies)
+        new_model = self._model_class.from_definition(definition, dependencies=dependencies_ids)
+
+        if not data_only:
+            self._create_model(new_model)
+
+        return new_model
+
+    def defined(self, match_criteria):
+        matching_definitions = self._select_definitions(match_criteria)
+        random.shuffle(matching_definitions)
+
+        for definition in matching_definitions:
+            model = self._create_from_definition(
+                definition=definition, new_dependencies=False, data_only=False
+            )
+            yield model
+
+    def new(self, match_criteria, new_dependencies=True, data_only=True):
+        matching_definitions = self._select_definitions(match_criteria)
```

Comment:
Should we do something when match_criteria do not match anything in configuration file?

Right now this will be an empty generator, so tests that want to handle a case when nothing matches need to do:

```
obj = next(data_provider.obj.new(...), None)
if not obj:
	# handle this gracefully
```

If test expects something to match and blindly uses generated value, it will fail with uncaught `StopIteration`. I think failing test is OK, but `StopIteration` is not very obvious failure root cause.

Reply:
Yeah, I think explicitly checking and handling `None` like this is better than letting a test blow up with `StopIteration`, as long as the checks we add will log/raise an error with more specific details.

Alternatively, if we expect this sanity-checking to be a common pattern, we could write a context manager to wrap the generator call, and that context manager could be responsible for checking `None` or excepting the `StopIteration` and also log an error or raise a more appropriate exception. I don't necessarily love this idea because it might get ugly or not work in all cases; I'm just dumping an idea here in case it inspires a better idea.

_Or_ the suggestion from my other comment about having a regular non-generator function/method for the use cases that don't need to iterate results.

Reply:
I think I'm just going to raise something inheriting from ValueError with a message along the lines of "provided match_criteria did not match anything; check your config file".

In [rare cases where it's ok-ish to not match anything](https://github.com/quipucords/camayoc/blob/master/camayoc/tests/qpc/cli/test_reports.py#L297), caller can just catch this exception.